### DISCUSSION TITLE
Anchor Placement Issues

### DISCUSSION THREAD

**Discussion by eunseo01** [bitcraze/discussions#614] [Q&A]
Hello, I am trying to fly multiple crazyflies using the Loco Positioning system. However, I keep getting the error 'ESTKALMAN: State out of bounds, resetting' in the console window of the cfclient. 
I have not been able to find the cause of this issue. Both the crazyflie and nodes are updated to the latest version, and eight anchors are placed in a rectangular space. 
Does the direction of the anchors affect the positioning of the crazyflies? 
I don't think it does, but I'm asking just to be sure. The attached picture shows the appearance of the anchors placed in the flying area. 
Please let me know if there are any issues related to the anchors. 
Thank you.

(Additionally, the width of the flying area is 560cm horizontally, 620cm vertically, and 240cm in height, and there are no obstacles around it. Even when I place the crazyflie in the center of the flying area and connect it to cfclient, it fails to determine its own position. All eight anchors, numbered 0 to 7, are displayed as being located at the center and remain stationary.)
![KakaoTalk_20230309_021049423](https://user-images.githubusercontent.com/107001681/223783577-f8a954a6-70aa-4702-bc66-141bca7f4ff5.jpg)
![KakaoTalk_20230309_021049423_01](https://user-images.githubusercontent.com/107001681/223783627-4a1d73b9-014b-4983-8ba1-d8616242f24d.jpg)

**Comment by krichardsson** [comment]
Hi!

It seems as the anchor positions have not been configured. You can find instructions of how to do it in [the Loco positioning system tutorial](https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/#enter-anchor-positions).

I'll give you an overview of how this works for completeness. 
The anchor positions are stored in the anchors and are transmitted in all UWB packets (regardless of mode). The estimator in the Crazyflie firmware uses the position data together with timing data from the UWB packets to estimate the position. Finally there is a memory mapping functionality to transfer the anchor position data to the client for visualization.

In your case I guess the anchor positions have not been configured and are set to the default (0, 0, 0).

When you set your anchor positions, the data will be transferred in the following order client -> Crazyflie -> anchors -> Crazyflie -> client, which means the update in the client will take a while.

> Does the direction of the anchors affect the positioning of the crazyflies?

You are right, that does not matter that much. The antennas have a radiation pattern that might make the estimate slightly better if they face towards the flying space, but it should not make much of a difference.

**Reply by eunseo01** [reply]
Thank you for your response. Is the ID of each anchor and the order of their placement important? Also, I'm curious if the x, y, z axes shown in the LPS tab need to match the actual orientation of the Crazyflie. Or is it okay for us to arbitrarily define the x, y, z axes and place the anchors accordingly?

**Reply by krichardsson** [reply]
> Is the ID of each anchor and the order of their placement important?

In TWR and TDoA2 you must use IDs 0-7, for TDoA3 any IDs can be used. 
In TDoA2 the anchors are transmitting in order (0, 1, 2 and so on) which in theory could make a difference of where anchors are placed, but I don't think it is noticable. I would recommend to use TDoA3 instead that is more robust (if you are interested in TDoA solutions).

> Also, I'm curious if the x, y, z axes shown in the LPS tab need to match the actual orientation of the Crazyflie. Or is it okay for us to arbitrarily define the x, y, z axes and place the anchors accordingly?

The Crazyflie must take off facing the positive X-axis as the Loco system can not measure yaw directly. The coordinate system is defined by where you place the anchors and the positions you configure in the anchors. The X and Y axis must be parallel to the floor and the Z axis pointing up, but apart from that you can define it as you like. There is an assumption though that the system is defined in such a way that the origin is in or close to the system, as the estimator will be reset to (0, 0, 0) by default. If you configure your origin "far away", the estimator might not converge.

**Reply by eunseo01** [reply]
In the Plotter tab, the values of ranging.distance are displayed as shown in the picture. Is there any problem with this? Also, in the Pose tab, the values of x and y are messy. They increase to a non-zero value and then return to the origin repeatedly. I tried changing the x-axis direction of the Crazyflie to the center of the flight space where the anchors were placed, but it had no effect. As shown in the stabilizer tab, the roll value also seems to be abnormal. Can this problem be solved? If there is any additional information needed to solve this problem, please let me know
![20230317_164508](https://user-images.githubusercontent.com/107001681/225846549-3d7ad3d7-6e8a-4f36-9ee3-b54dc012d182.png)
![20230317_164551](https://user-images.githubusercontent.com/107001681/225846556-e1f6e8f7-c288-44c8-a54c-e10c34a39cba.png)
![20230317_164638](https://user-images.githubusercontent.com/107001681/225846557-cdabf706-8f4c-41ea-9581-2ef5f81820a5.png)
![20230317_164658](https://user-images.githubusercontent.com/107001681/225846560-0701ba5f-3fb9-4570-b329-60313a0e8c4f.png)
![20230317_164731](https://user-images.githubusercontent.com/107001681/225846561-e98d3909-0ccd-48d4-9d22-3d8654174aa8.png)
![20230317_164805](https://user-images.githubusercontent.com/107001681/225846563-e031e702-2565-4484-a96d-10a6858a65ce.png)

**Reply by krichardsson** [reply]
The problem is probably not the ranging, it seems to be good. As mentioned earlier I think the problem is that the anchor positions have not been configured. 
When you have configured the anchor positions properly, the anchors will be displayed in their positions in the 3D graph in the Loco tab, similar to what you can [see in the documentation](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/loco_tab/#position-estimate-mode). Do you get this?

**Reply by eunseo01** [reply]
Follow the link below 
(https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/#anchor-positions)
I reinstalled 8 anchors as shown in the following photo. (Width: 6m / Length: 3m / Height: 3m)
But it still doesn't solve the problem of crazyflie not finding its location. Is there another reason other than the anchor placement problem?

Also, I wonder if there is a way to directly adjust the coordinate system.

![image](https://user-images.githubusercontent.com/107001681/227695731-be096f97-bd2d-484a-89da-1c0462b624a5.png)

**Reply by krichardsson** [reply]
Just to clarify, I'm not talking about the physical position of the anchors, they can be placed anywhere (sort of). I mean that you have physically measure where the anchors are placed and pass this information to the Crazyflie by [configuring the anchor positions in the client](https://www.bitcraze.io/documentation/tutorials/getting-started-with-loco-positioning-system/#enter-anchor-positions). When you have configured the positions in the client you should see the anchors in the 3D graph. 

> Also, I wonder if there is a way to directly adjust the coordinate system.

You can decide freely where you want your origin to be located, measure where the anchors are in your coordinate system and enter the positions in the anchor configuration (mentioned above).

**Reply by eunseo01** [reply]
As you suggested, I manually entered the coordinates of the anchor and it appeared correctly in the LPS tab, and the Crazyflie successfully completed stable flight! I sincerely appreciate your help. 
I have one question to ask. Is it normal that the coordinates of the anchor are not automatically displayed, and we have to manually measure and enter them?

**Reply by krichardsson** [reply]
Great, I'm happy it works!

> I have one question to ask. Is it normal that the coordinates of the anchor are not automatically displayed, and we have to manually measure and enter them? 

Yes, it is the normal behavior. There has been some attempts at automatically estimating the positions of the anchors, but we have not found the solutions to be good enough.
